### PR TITLE
userspace/su: implement basic features

### PR DESCRIPTION
**Pull request by Scirese**
If we want to get KernelSU version by `CMD_GET_VERSION`, `root` must be allowed to be manager.

Thanks Ylarod for assistance!

### CODE REVIEW

**Comment by tiann** [comment]
And also, I think we should keep the original `su.c`, it is simple enough to use and easy enough to understand. Maybe rename your implementation to `su2.c` or something else?

**Comment by Scirese** [comment]
> And also, I think we should keep the original `su.c`, it is simple enough to use and easy enough to understand. Maybe rename your implementation to `su2.c` or something else?

I think it's better to replace the original `su` with this one and move the orignal one to `su2.c`.
Because this implementation has better compactibilty with other popular `su` implementation like `MagiskSU`.

**Comment by tiann** [comment]
> > And also, I think we should keep the original `su.c`, it is simple enough to use and easy enough to understand. Maybe rename your implementation to `su2.c` or something else?
> 
> I think it's better to replace the original `su` with this one and move the orignal one to `su2.c`. Because this implementation has better compatibilty with other popular `su` implementation like `MagiskSU`.

rename to original one to `simple_su.c` is ok

**Comment by Scirese** [comment]
If no other problems, the change is ready to be reviewed.

**Review by Ylarod — CHANGES_REQUESTED** [review]
```
include $(CLEAR_VARS)
LOCAL_MODULE := su
LOCAL_SRC_FILES := su.c
include $(BUILD_EXECUTABLE)

include $(CLEAR_VARS)
LOCAL_MODULE := su2
LOCAL_SRC_FILES := simple_su.c
include $(BUILD_EXECUTABLE)
```

File: kernel/ksu.c [hunk 1]
```diff
@@ -89,7 +89,13 @@ static uid_t __manager_uid;
 
 static bool is_manager()
 {
-	return __manager_uid == current_uid().val;
+	if ( __manager_uid == current_uid().val ) {
```
Comment (Ylarod) [inline kernel/ksu.c:92]: I think using `||` is better

File: userspace/su/jni/Android.mk [hunk 2]
```diff
@@ -1,7 +1,7 @@
 LOCAL_PATH := $(call my-dir)
 
 include $(CLEAR_VARS)
-LOCAL_MODULE := su
```
Comment (Ylarod) [inline userspace/su/jni/Android.mk:4]:
```
include $(CLEAR_VARS)
LOCAL_MODULE := su
LOCAL_SRC_FILES := su.c
include $(BUILD_EXECUTABLE)

include $(CLEAR_VARS)
LOCAL_MODULE := su2
LOCAL_SRC_FILES := simple_su.c
include $(BUILD_EXECUTABLE)
```

File: userspace/su/jni/su.c [hunk 3]
```diff
@@ -1,10 +1,118 @@
...
+#define CMD_GRANT_ROOT 0
+
+#define CMD_BECOME_MANAGER 1
+#define CMD_GET_VERSION 2
+#define CMD_ALLOW_SU 3
+#define CMD_DENY_SU 4
+#define CMD_GET_ALLOW_LIST 5
+#define CMD_GET_DENY_LIST 6
+
+#define VERSION "8"
```
Comment (Ylarod) [inline userspace/su/jni/su.c:17]: Get version from prctl

**Review by Ylarod — CHANGES_REQUESTED** [review]

File: userspace/su/jni/su.c (hunk 3 above)
Comment (Scirese) [reply userspace/su/jni/su.c:17]: Sorry, I just forgot to remove this macro. It already use prctl to get version.

File: userspace/su/jni/su.c [hunk 4]
```diff
@@ -1,10 +1,118 @@
...
+}
+
+int main(int argc, char *argv[], char *envp[]){
+
+    if(argc == 1) {
+        su_shell(envp);
+    }
+
+    int opt;
+    int version = getver();
```
Comment (Ylarod) [inline userspace/su/jni/su.c:99]: We don't need to get version if `-v` is not existed.
```
case 'v': {
    int version = getver();
    printf("%d:KernelSU\n", version );
     exit(EXIT_SUCCESS);
}
```

File: kernel/ksu.c (hunk 1 above)
Comment (Scirese) [reply kernel/ksu.c:92]:
```
	if ( __manager_uid || 0 == current_uid().val ) {
		return true;
	} else {
		return false;
	}
```
Like this?

Comment (Ylarod) [reply kernel/ksu.c:92]:
```
return __manager_uid == current_uid().val || 0 == current_uid().val;
```

**Review by Ylarod — APPROVED** [review]

File: kernel/ksu.c [hunk 5]
```diff
@@ -89,7 +89,7 @@ static uid_t __manager_uid;
 
 static bool is_manager()
 {
-	return __manager_uid == current_uid().val;
+	return __manager_uid == current_uid().val || 0 == current_uid().val;
```
Comment (tiann) [inline kernel/ksu.c:92]: we should not allow any root process to be manager, if you want to get verion, just allow root process to get version, avoiding widen the scope.

Comment (tiann) [reply kernel/ksu.c:92]: And also, you should put this to another PR, don't mess it up.

**Review by tiann — CHANGES_REQUESTED** [review]

File: kernel/ksu.c (hunk 5 above)
Comment (Scirese) [reply kernel/ksu.c:92]: If don't allow this, we can:
Allow non-manager to excute `CMD_GET_VERSION`;
Hardcode it
which is better？

Comment (Scirese) [reply kernel/ksu.c:92]: I think this is not a breaking change, so I didn't put it into another PR.

Comment (Ylarod) [reply kernel/ksu.c:92]: Each time you push commits to this branch, the actions will compile the whole kernel for 9 times.
I've canceled action runs for many times.

File: userspace/su/jni/su.c [hunk 6]
```diff
@@ -1,10 +1,118 @@
...
+    return result == 0xDEADBEEF;
+
+}
+
+void elevate(){
+
+    // Talk to Daemon in Kernel Space
+    bool status = ksuctl(CMD_GRANT_ROOT, 0, NULL);
+
+    if(status){
```
[the remark this answers is not in the document]
Comment (Scirese) [reply userspace/su/jni/su.c:36]: I didn't see where is wrong, so could you tell me what should I do?
Sorry for my poor C skill...

File: kernel/ksu.c (hunk 5 above)
Comment (Scirese) [reply kernel/ksu.c:92]: > Each time you push commits to this branch, the actions will compile the whole kernel for 9 times. I've canceled action runs for many times.

Sorry for that

Comment (tiann) [reply kernel/ksu.c:92]: > I think this is not a breaking change, so I didn't put it into another PR.

This IS a breaking change.

Comment (tiann) [reply kernel/ksu.c:92]: You just need to get the version, why allow all root process do manager operation?

Comment (Scirese) [reply kernel/ksu.c:92]: > You just need to get the version, why allow all root process do manager operation?

My fault, sorry.

**Comment by Scirese** [comment]
Ok, solved

File: userspace/su/jni/su.c [hunk 7]
```diff
@@ -1,10 +1,113 @@
...
+    }
+
+}
+
+int getver() {
+
+    elevate();
+
+    int32_t version = -1;
+    if (ksuctl(CMD_GET_VERSION, & version, NULL)) {
```
Comment (tiann) [inline userspace/su/jni/su.c:49]: delete `if`

File: userspace/su/jni/su.c [hunk 8]
```diff
@@ -1,10 +1,113 @@
...
+
+void su_command(char * command[], int len) {
+
+    elevate();
+    int status;
+
+    if (len == 1) {
+        status = system(command[0]);
+    } else {
+        command[len] = NULL;
```
Comment (tiann) [inline userspace/su/jni/su.c:80]: Is this necessary? the args passed by `main` is already null-terminated(maybe you should check it)?

**Review by tiann — CHANGES_REQUESTED** [review]

**Comment by tiann** [comment]
please check the review, it would be ok if it is resolved : )

File: userspace/su/jni/su.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiann) [reply userspace/su/jni/su.c]:
```
if(status){
```

should be:

```
if (status) {
```
(the space before and after `if)

By the way, you can use a code format plugin.

File: userspace/su/jni/su.c (hunk 8 above)
Comment (Scirese) [reply userspace/su/jni/su.c:80]: Seems works without it, I will delete it.

**Comment by Scirese** [comment]
OK

File: userspace/su/jni/su.c [hunk 9]
```diff
@@ -1,10 +1,110 @@
...
+
+}
+
+void elevate() {
+
+    // Talk to Daemon in Kernel Space
+    bool status = ksuctl(CMD_GRANT_ROOT, 0, NULL);
+
+    if (status) {
+        fprintf(stderr, "Permission denied\n");
```
[the remark this answers is not in the document]
Comment (tiann) [reply userspace/su/jni/su.c:38]: Is this resolved?

**Review by tiann — CHANGES_REQUESTED** [review]

File: userspace/su/jni/su.c [hunk 10]
```diff
@@ -1,10 +1,110 @@
...
+
+}
+
+static bool ksuctl(int cmd, void * arg1, void * arg2) {
+
     int32_t result = 0;
-    prctl(0xdeadbeef, 0, 0, 0, &result);
-    system("/system/bin/sh");
+    prctl(0xDEADBEEF, cmd, arg1, arg2, & result);
+    return result == 0xDEADBEEF;
```
[the remark this answers is not in the document]
Comment (Scirese) [reply userspace/su/jni/su.c:28]: I copy this from the manager.

File: userspace/su/jni/su.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (Scirese) [reply userspace/su/jni/su.c:38]: Yes, There is a error at here

**Comment by Scirese** [comment]
Ready, Sorry for the inconivence I brought

**Comment by tiann** [comment]
No need to apologize, we are just working together!

**Review by tiann — APPROVED** [review]

File: userspace/su/jni/Android.mk [hunk 11]
```diff
@@ -3,5 +3,9 @@ LOCAL_PATH := $(call my-dir)
 include $(CLEAR_VARS)
 LOCAL_MODULE := su
 LOCAL_SRC_FILES := su.c
+include $(BUILD_EXECUTABLE)
 
+include $(CLEAR_VARS)
+LOCAL_MODULE := su2
```
Comment (tiann) [inline userspace/su/jni/Android.mk:9]: I checked the CI result and i think we should change the module name to `simple_su`, `su` and `su2` is confusing!

**Review by tiann — CHANGES_REQUESTED** [review]

**Review by tiann — APPROVED** [review]

**Pull request merged**